File: Medium/3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring(SlidingWindow).cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int left = 0;
        int right = 0;
        int maxLen = 0;
        unordered_map<char, int> window;
        while (right < n) {
            char c1 = s[right];
            window[c1]++;
            right++;
            while (window[c1] > 1) {
                char c2 = s[left];
                window[c2]--;
                left++;
            }
            maxLen = max(maxLen, right - left);
        }
        return maxLen;
    }
};

File: Medium/3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring(sliding window).java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l = 0;
        int r = 0;
        Map<Character, Integer> window = new HashMap<>();
        int ans = 0;
        while(r < s.length()){
            char c1 = s.charAt(r);
            window.put(c1, window.getOrDefault(c1, 0) + 1);
            r++;
            while(window.get(c1) > 1){
                char c2 = s.charAt(l);
                window.put(c2, window.get(c2) - 1);
                l++;
            }
            ans = Math.max(ans, r - l);
        }
        return ans;
    }
}